### ISSUE TITLE
v2: cli output info

### ISSUE THREAD

**Issue by bartrail** [aliparlakci/bulk-downloader-for-reddit#261]
I have just recently tested v2 and like the overall progress very much, also the refactoring looks really promising :-) Thanks for continuing the development!

I would like to ask if you are planning to put some of the outputs that v1.9 had in v2 as well:

- Getting Posts
- Download progress for every entry 
  - Overall Progress ( 45 / 294 )
  - Album Progress ( 5 / 10 )
  - Downloaded MB ( 5 / 10MB ) 
- Filename while Downloading
- End Result: 
- There were 4 duplicates
  Total of 230 links downloaded!

Thanks!

**Comment by aliparlakci** [comment]
`--verbose` parameter gives more information on the console log. You might want to try it.

If it is not enough for you, can you state why do you need those information?

**Comment by bartrail** [comment]
You're right, `--verbose` gives the filename and if the file already exists :)

I think it is a useful information to show the current state / progress of a download like on every other application (and before) as well. 
If you start a download and you abort it, it is nice to know how far you've come. Maybe also how much data was downloaded (at the end)

**Comment by aliparlakci** [comment]
But if you abort it, you already lose all the progress. Because the file lives inside memory until it is fully downloaded. It will only make you suffer from the regret aborting the download :)

I still don't understand why this information is beneficial.

**Comment by bartrail** [comment]
> But if you abort it, you already lose all the progress. Because the file lives inside memory until it is fully downloaded. It will only make you suffer from the regret aborting the download :)

We are talking about two kinds of progresses here:
1. Overall progress of the download (like file 100 / 1000) 
2. By Byte progress of a file while being downloaded.

To 1:
I think it is very useful to know when I aborted or how many files will be left. I actually don't see the arguments why you removed this feature

To 2:
I see your point that the file is removed when I abort a download and only those which are 100% finished stay. Of course. But sometimes gfycat API is quite slow or someone uploaded a > 100MB mp4 - I just think it is quite useful to see the file size and the progress of it while being downloaded. As we have a `--exclude-id` option in v2 I would use this information to skip a file if it is slow or keeps crashing with a timeout.

**Comment by aliparlakci** [comment]
For the first one, bdfr no longer fetches all the posts before downloading. So, how many files are left is a mystery. Otherwise, it would have to wait for some time before starting the downloading and it is not useful.

For the second one, I want to hear what @Serene-Arc would say.

**Comment by aliparlakci** [comment]
I want to add more to first one. Reddit only gives 100 posts at a time. To download 1000 posts, there needs to be 10 requests to reddit and they cannot be in parallel because we need the information in the previous request to make the new request. Unfortunately, it takes ages for reddit to respond. Considering bdfr being able to download tens of thousands of posts, making all the requests one after other takes too much time. That is why we don't fetch all the posts first.

**Comment by bartrail** [comment]
Yes, I recognized the `GETTING POSTS` from v1 and the dots. I see why you dropped that to increase the overall performance. But maybe this could be a configuration option to prefetch all posts before so a total download progress would be possible again?

**Comment by Serene-Arc** [comment]
It's a complicated issue. 

For the first option, it would be relatively easy to do with tqdm, which is a Python progress bar library. But it looks ugly as hell when used with logging like we are here. My view up until now has been that, if we do include something like that, then it would be for the explicit purpose of informing a user if they are actively staring at the screen. This would likely be the case with a ncurses interface similar to v1 where there is an interactive menu. My design philosophy up until now in the project has been that this is a core that can function headlessly in a server as a script, or from the commandline. It's informative, but not geared towards active user interaction, which is what a progress bar would be for.

Actually we do know how many submissions there are total to download fairly quickly, just not how many are valid and will actually download.

Now, for the second option, there is a difficulty. To show the process of actively downloading a resource, we would again need to interject with the flow of logging data regularly, which looks awful and makes everything much much much harder to read. Additionally, there is, regretfully, some disparity between how and when resources are actually downloaded. For some resources, this might be technically possible, but not for others. See, for most downloader modules, they merely return the link to a resource which is then downloaded. However, several downloaders, most notably the youtube and children modules, actually download first, much earlier than the other modules. This is because there is no direct link that we can use to download, so we must use a third party application, like youtube-dl, to actually get it. So to show *that* download would be in a completely different part of the program to the rest of them. If you only show the download in one place, then a youtube download will appear to complete instantly, while the parsing of submissions will take longer, since that's where the download actually takes place.

Side note: if you want a look of what logging and tqdm look like together, check out the podcast downloader on my profile. Long story short, it looks awful and it's likely something that I'll remove soon

**Comment by Serene-Arc** [comment]
All submissions are fetched very quickly, but that's really not a big part of the program. it's parsing them for the proper resource, and then downloading the actual resource itself.

**Comment by bartrail** [comment]
Okay I see. Thanks for explaining and I see that showing a progress is difficult and looks awful if you pipe it through the regular log stream (especially when viewed later in a static log file)

Then maybe I can ask for the last information that v1.9 had: The total amount of downloaded files and how many of those were duplicates? :-)

**Comment by Serene-Arc** [comment]
Well, that's pretty easy to calculate yourself from the log file.

For the amount of files downloaded in the last run, try:

```bash
cat log_output.txt | grep -i 'downloaded submission' | wc -l
```

For the number of duplicates that were found:

```bash
cat log_output.txt | grep 'already exists, continuing' | wc -l
```

Would these be sufficient, or do you want some summary or something at the end of the program at the time of completion?

**Comment by bartrail** [comment]
I already thought about grepping the output, but I'd rather see it in the written logfile itself ;-) 
but surely this is just a nice to have feature. 

As already said in my first comment here, I just noticed the differences in output and wanted to know if you are going to keep them or not ;-)

**Comment by Serene-Arc** [comment]
I wasn't planning to do it here, no. I feel summary statistics wouldn't really be much of a useful addition to the code. Maybe an interactive version of the program could make use of it, but currently, it would be a simple counter inside the download loop, which would increment every time that logging message is printed anyway. They'd have the same result.

So in short, no.
